File: app/src/main/java/com/nfcalarmclock/NacNameDialog.java
package com.nfcalarmclock;

import android.app.AlertDialog;
import android.content.Context;
import android.content.res.ColorStateList;
import android.graphics.Color;
import android.view.KeyEvent;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.widget.TextView;
import androidx.core.graphics.ColorUtils;
import com.google.android.material.textfield.TextInputLayout;
import com.google.android.material.textfield.TextInputEditText;

/**
 * The dialog class that will handle saving the name of the alarm.
 */
public class NacNameDialog
	extends NacDialog
	implements TextView.OnEditorActionListener,
		NacDialog.OnDismissListener,
		NacDialog.OnShowListener
{

	/**
	 */
	public NacNameDialog()
	{
		super(R.layout.dlg_alarm_name);

		addOnDismissListener(this);
		addOnShowListener(this);
	}

	/**
	 * Build the dialog.
	 */
	@Override
	public void onBuildDialog(Context context, AlertDialog.Builder builder)
	{
		NacSharedConstants cons = new NacSharedConstants(context);

		//builder.setTitle(cons.getTitleName());
		setPositiveButton(cons.getActionOk());
		setNegativeButton(cons.getActionCancel());
	}

	/**
	 * Save the alarm name as the dialog data.
	 */
	@Override
	public boolean onDismissDialog(NacDialog dialog)
	{
		View root = dialog.getRoot();
		TextInputEditText editText = root.findViewById(R.id.name_entry);
		String name = editText.getText().toString();

		dialog.saveData(name);
		return true;
	}

	/**
	 * Close the keyboard when the user hits enter.
	 */
	@Override
	public boolean onEditorAction(TextView tv, int action, KeyEvent event)
	{
		if ((event == null) && (action == EditorInfo.IME_ACTION_DONE))
		{
			dismiss();
			return true;
		}
		return false;
	}

	/**
	 * Setup the views when the dialog is shown.
	 */
	@Override
	public void onShowDialog(NacDialog dialog, View root)
	{
		Context context = root.getContext();
		NacSharedPreferences shared = new NacSharedPreferences(context);
		String name = this.getDataString();
		int theme = shared.getThemeColor();
		int blendedTheme = ColorUtils.blendARGB(theme, Color.TRANSPARENT, 0.1f);
		ColorStateList colorStateList = ColorStateList.valueOf(blendedTheme);
		TextInputLayout editBox = root.findViewById(R.id.name_box);
		TextInputEditText editText = root.findViewById(R.id.name_entry);

		editText.setText(name);
		editText.selectAll();
		editText.setOnEditorActionListener(this);
		editText.setImeOptions(EditorInfo.IME_ACTION_DONE);
		//editText.setBackgroundTintList(ColorStateList.valueOf(themeColor));
		editBox.setBoxStrokeColor(blendedTheme);
		editBox.setHintTextColor(colorStateList);
		showKeyboard();
		scale(0.85, 0.5, false, true);
	}

}

File: app/src/main/kotlin/com/nfcalarmclock/alarm/options/name/NacNameDialog.kt
package com.nfcalarmclock.alarm.options.name

import android.app.AlertDialog
import android.app.Dialog
import android.content.Context
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.view.inputmethod.EditorInfo
import android.view.inputmethod.InputMethodManager
import com.google.android.material.textfield.TextInputEditText
import com.google.android.material.textfield.TextInputLayout
import com.nfcalarmclock.R
import com.nfcalarmclock.view.dialog.NacDialogFragment
import com.nfcalarmclock.view.setupInputLayoutColor

/**
 * The dialog class that will handle saving the name of the alarm.
 */
class NacNameDialog
	: NacDialogFragment()
{

	/**
	 * Listener for when a name is entered.
	 */
	fun interface OnNameEnteredListener
	{
		fun onNameEntered(name: String)
	}

	/**
	 * Default alarm name to show.
	 *
	 * This will be set externally before the dialog is shown.
	 */
	var defaultName : String = ""

	/**
	 * Listener for when the name is entered.
	 */
	var onNameEnteredListener: OnNameEnteredListener? = null

	/**
	 * Called when the dialog is created.
	 */
	override fun onCreateDialog(savedInstanceState: Bundle?): Dialog
	{
		// Setup the shared preferences
		setupSharedPreferences()

		// Create the dialog
		return AlertDialog.Builder(requireContext())
			.setPositiveButton(R.string.action_ok) { _, _ ->

				// Get the name that is in the edit text
				val editText = dialog!!.findViewById<TextInputEditText>(R.id.name_entry)
				val name = editText.text.toString()

				// Call the listener
				onNameEnteredListener?.onNameEntered(name)

			}
			.setNegativeButton(R.string.action_cancel, null)
			.setView(R.layout.dlg_alarm_name)
			.create()
	}

	/**
	 * Called when the fragment is resumed.
	 */
	override fun onResume()
	{
		// Super
		super.onResume()

		// Get the context
		val context = requireContext()

		// Get the views
		val inputLayout: TextInputLayout = dialog!!.findViewById(R.id.name_box)
		val editText: TextInputEditText = dialog!!.findViewById(R.id.name_entry)

		// Set the dialog background color
		dialog!!.window?.setBackgroundDrawableResource(R.color.gray)

		// Set text to show
		editText.setText(defaultName)

		// Select all the text so that the user can delete it easily if they
		// choose to
		editText.selectAll()

		// This will show the newline button in the soft keyboard, since the
		// text entry is multi-line
		editText.imeOptions = EditorInfo.IME_ACTION_DONE

		// Setup the color
		inputLayout.setupInputLayoutColor(context, sharedPreferences!!)

		// Show the keyboard. This occurs on a delay so that the window has time to
		// be in focus
		waitToShowKeyboard(editText)
	}

	/**
	 * Show the keyboard.
	 */
	@Suppress("deprecation")
	private fun showKeyboard(editText: TextInputEditText)
	{
		// Define the context
		val context: Context

		try
		{
			// Get the context
			context = requireContext()
		}
		catch (_: IllegalStateException)
		{
			return
		}

		// Request focus
		editText.requestFocus()

		// Check API >= 31
		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S)
		{
			val imm = context.getSystemService(InputMethodManager::class.java)

			// Show implicit
			imm.showSoftInput(editText, InputMethodManager.SHOW_IMPLICIT)
		}
		// API < 31
		else
		{
			val imm = context.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager

			// Show forced
			imm.toggleSoftInput(InputMethodManager.SHOW_FORCED, 0)
		}
	}

	/**
	 * Wait to show the soft keyboard.
	 */
	private fun waitToShowKeyboard(editText: TextInputEditText)
	{
		val looper = requireContext().mainLooper
		val handler = Handler(looper)

		// Wait for 200 ms and then show the keyboard
		handler.postDelayed({
			showKeyboard(editText)
		}, 200)
	}

	companion object
	{

		/**
		 * Tag for the class.
		 */
		const val TAG = "NacNameDialog"

		/**
		 * Create a dialog that can be shown easily.
		 */
		fun create(
			name: String,
			onNameEnteredListener: (String) -> Unit = {}
		): NacNameDialog
		{
			// Create the dialog
			val dialog = NacNameDialog()

			// Set the default name
			dialog.defaultName = name

			// Set the listener
			dialog.onNameEnteredListener = OnNameEnteredListener { n ->
				onNameEnteredListener(n)
			}

			return dialog
		}

	}

}